### ISSUE TITLE
"No POST data" in error report

### ISSUE THREAD

**Issue by bblanchon** [encode/django-rest-framework#7953]
## Checklist

- [X] Raised initially as discussion #7899
- [X] This cannot be dealt with as a third party library. (We prefer new functionality to be [in the form of third party libraries](https://www.django-rest-framework.org/community/third-party-packages/#about-third-party-packages) where possible.)
- [X] I have reduced the issue to the simplest possible case.

## Description

I'm using Django 3.1.7 & DRF 3.12.2.
When an exception occurs in a `ViewSet`, Django automatically sends an email to the administrator.

The problem is that the **POST data is always empty** in this email:

> POST: No POST data

Instead, **I expect to see the JSON payload** that was sent in the request body.

I added a breakpoint in `django.utils.log.AdminEmailHandler.emit()`, and I can confirm that `request.POST` is empty.
But on the other hand, I'm sure that `request.data` was not empty in the `ViewSet`.

I verified that the request's `Content-Type` and `X-Requested-With` are correctly set.
The only non-standard middlewares I use are [corsheaders](https://pypi.org/project/django-cors-headers/) and [admin_reorder](https://pypi.org/project/django-modeladmin-reorder/); I don't think they are related to this problem.

**Comment by danielmoessner** [comment]
Any progress on this issue? I have the same problem.

**Comment by Hafnernuss** [comment]
I can also confirm this issue, and it makes it a little bit hard to trace back errors

**Comment by rvinzent** [comment]
At this point I am convinced DRF is a dead project

**Comment by bblanchon** [comment]
I still think this is an important feature, please don't close this issue.

**Comment by Hafnernuss** [comment]
I agree.

**Comment by markdawson** [comment]
I'm also having this issue.

**Comment by bblanchon** [comment]
I still think this is an important feature, please don't close this issue.

**Comment by donghj2000** [comment]
It is because django dose not parse body of application/json.You can convert json string to dict  by yourself, like this:
res = request.body
json_str = res.decode('utf-8')   
json_dict = json.loads(json_str)   
Then write a custom log handler class to send your own dict data converted from json.

**Comment by donghj2000** [comment]
def emit(self, record):
        try:
            request = record.request
            subject = '%s (%s IP): %s' % (
                record.levelname,
                ('internal' if request.META.get('REMOTE_ADDR') in settings.INTERNAL_IPS
                 else 'EXTERNAL'),
                record.getMessage()
            )

            import json
            from django.http import QueryDict
            json_str = request.body.decode('utf-8')
            json_dict = json.loads(json_str)
            query_dict = QueryDict(mutable=True)
            for key in json_dict.keys():
                query_dict[key] = json_dict[key]
            request.POST = query_dict
            
        except Exception as ex:
            subject = '%s: %s' % (
                record.levelname,
                record.getMessage()
            )
            request = None
        subject = self.format_subject(subject)

**Comment by donghj2000** [comment]
some code snippet:
```
 def emit(self, record):
        try:
            request = record.request
            subject = '%s (%s IP): %s' % (
                record.levelname,
                ('internal' if request.META.get('REMOTE_ADDR') in settings.INTERNAL_IPS
                 else 'EXTERNAL'),
                record.getMessage()
            )

            import json
            from django.http import QueryDict
            json_str = request.body.decode('utf-8')
            json_dict = json.loads(json_str)
            query_dict = QueryDict(mutable=True)
            for key in json_dict.keys():
                query_dict[key] = json_dict[key]
            request.POST = query_dict
            
        except Exception as ex:
            subject = '%s: %s' % (
                record.levelname,
                record.getMessage()
            )
            request = None
        subject = self.format_subject(subject)
```

**Comment by donghj2000** [comment]
Glad to help you. I have not found why they didn't do so. Another workaround is to use Middleware:
```
class ParseJsonData(MiddlewareMixin):
    def process_request(self, request):
        if request.content_type.startswith("application/json"):
            try:
                json_str = request.body.decode('utf-8')
                json_dict = json.loads(json_str)
                query_dict = QueryDict(mutable=True)
                for key in json_dict.keys():
                    query_dict[key] = json_dict[key]
                request.POST = query_dict
            except Exception as ex:
                pass
```

**Comment by rvinzent** [comment]
> Glad to help you. I have not found why they didn't do so. Another workaround is to use Middleware:
> ```
> class ParseJsonData(MiddlewareMixin):
>     def process_request(self, request):
>         if request.content_type.startswith("application/json"):
>             try:
>                 json_str = request.body.decode('utf-8')
>                 json_dict = json.loads(json_str)
>                 query_dict = QueryDict(mutable=True)
>                 for key in json_dict.keys():
>                     query_dict[key] = json_dict[key]
>                 request.POST = query_dict
>             except Exception as ex:
>                 pass
> ```

This snippet has an issue where it assumes the JSON is an object and will fail on any other valid JSON value and will presumably bubble up 500s for invalid JSON bodies when a 400 would be the proper response status.

It's a fine work around but folks here should probably avoid copy pasting that for any production use case.

**Comment by donghj2000** [comment]
It seems to be a bug exactly, but request.POST is  initialized as a QueryDict(django.http.request.py->class HttpRequest) ,so I don't know how to deal with other JSON type...

**Comment by auvipy** [comment]
> At this point I am convinced DRF is a dead project

no. it has 4 new more maintainers now. btw would love to see a failing test case to find out the fix for the problem. who is up for that?

**Comment by jonnythebard** [comment]
@auvipy 
> would love to see a failing test case to find out the fix for the problem

I can do this but I'd like to know the definition of the problem from your side.

@donghj2000 said,
> It is because django dose not parse body of application/json to `request.POST` field.

Do you agree that it is a problem and `request.POST` also has to parse application/json?
I am asking this because I thought there must be a DRF design reason behind `request.POST` is not parsing application/json.
If you conceive this to be a bug rather than a DRF design, I can raise a fix PR.

**Comment by WhiteSage** [comment]
The problem seems to be that the DRF parsers consume Django's HttpRequest without setting its POST attribute, and return a DRF Request class instance. When Django composes the error message DRF's request instance is no longer available, and has not post data in its own HttpRequest. This could be solved by overriding initialize_request on the APIView (or ViewSet) class:

```
    def initialize_request(self, request, *args, **kwargs):
        drf_request= super().initialize_request(request, *args, **kwargs)
        request.POST = drf_request.data
        return DRFRequest
```

I would like to know, however, if not setting the POST attribute is by design, and what is the rationale behind it. I'm usually wary of changing the default behavior, but it is clearly incompatible with Django's logging system.

**Comment by sevdog** [comment]
I was looking at this problem and caught #5800 which was the point in which the `if` which prevent the setting of files/data to be propagated to the original request.

That PR refers to a bug introduced in #5590 (which was without that condition), however there are no clue about why this should be a bug (since here we are complaining that the absence of it is a bug):

> request POST/FILES were also set for requests that did not have a form media type

**Comment by zainnadeem786** [comment]
Hi @auvipy,

I'd be interested in investigating this further.

Before proposing a fix, I'd like to verify the current behavior around #5590 and #5800 and determine whether the expected outcome is:

1. Propagating parsed JSON data back to the underlying Django request.POST for error reporting purposes, or
2. Preserving the current separation between request.POST and DRF request.data and addressing the admin email use case differently.

I'll try to reduce this to a failing test case first and report back with findings.

Thanks.

**Comment by zainnadeem786** [comment]
Hi @auvipy,

I investigated the current behavior on main and was able to reproduce the report.

From what I found, DRF correctly parses JSON into `request.data`, while Django's original `HttpRequest.POST` remains empty. This appears to be intentional rather than a parsing bug.

Looking through the history, #5590 preserved the empty POST behavior for JSON requests and #5800 further restricted synchronization to form media types. There are also existing tests protecting the current behavior.

Because of that, I'm not sure a failing test asserting JSON data appears in `request.POST` would be aligned with the current design.

Before preparing a PR, could you clarify whether the preferred direction would be:

1. A DRF-aware error-reporting/reporter integration that exposes parsed request data without modifying `request.POST`, or
2. Documentation explaining the limitation and recommended customization points?

I'd be happy to continue from there once the expected behavior is confirmed.

**Comment by auvipy** [comment]
<i am open to a proper fix to this as this seems to be a regression, apparently
